Exercises/chapter7/e1-a-robot.js

```diff
@@ -0,0 +1,192 @@
+    addEdge(to, from);
+  }
+  return graph;
+}
+
+const roadGraph = buildGraph(roads);
+
+class VillageState {
+  constructor(place, parcels) {
+    this.place = place;
+    this.parcels = parcels;
+  }
+
+  move(destination) {
+    if (!roadGraph[this.place].includes(destination)) {
+      return this;
+    } else {
+      let parcels = this.parcels.map(p => {
+        if (p.place != this.place) return p;
+        return {place: destination, address: p.address};
+      }).filter(p => p.place != p.address);
+      return new VillageState(destination, parcels);
+    }
+  }
+
+  static random(parcelCount = 5) {
+    let parcels = [];
+    for (let i = 0; i < parcelCount; i++) {
+      let address = randomPick(Object.keys(roadGraph));
+      let place;
+      do {
+        place = randomPick(Object.keys(roadGraph));
+      } while (place === address);
+      parcels.push({place, address});
+    }
+    return new VillageState("Post Office", parcels);
+  };
+}
+
+function runRobot(state, robot, memory) {
+  for (let turn = 0;; turn++) {
+    if (state.parcels.length === 0) {
+      return turn;
+    }
+    let action = robot(state, memory);
+    state = state.move(action.direction);
+    memory = action.memory;
+
+    if (turn > 50) {
+      console.log(state);
+      return;
+    }
+    // console.log(state);
+    //console.log(`Moved to ${action.direction}`);
+  }
+}
+
+function randomPick(array) {
+  let choice = Math.floor(Math.random() * array.length);
+  return array[choice];
+}
+function findRoute(graph, from, to) {
+  let work = [{at: from, route: []}];
+  for (let i = 0; i < work.length; i++) {
+    let {at, route} = work[i];
+    for (let place of graph[at]) {
+      if (place === to) return route.concat(place);
+      if (!work.some(w => w.at === place)) {
+        work.push({at: place, route: route.concat(place)});
+      }
+    }
+  }
+}
+
+function randomRobot(state) {
+  return {direction: randomPick(roadGraph[state.place])};
+}
+function routeRobot(state, memory) {
+  const mailRoute = [
+    "Alice's House", "Cabin", "Alice's House", "Bob's House",
+    "Town Hall", "Daria's House", "Ernie's House",
+    "Grete's House", "Shop", "Grete's House", "Farm",
+    "Marketplace", "Post Office"
+  ];
+
+  if (memory.length === 0) {
+    memory = mailRoute;
+  }
+  return {direction: memory[0], memory: memory.slice(1)};
+}
+function goalOrientedRobot({place, parcels}, route) {
+  if (route.length === 0) {
+    let parcel = parcels[0];
+    if (parcel.place != place) {
+      route = findRoute(roadGraph, place, parcel.place);
+    } else {
+      route = findRoute(roadGraph, place, parcel.address);
+    }
+  }
+  return {direction: route[0], memory: route.slice(1)};
+}
+
+/**
+  This robot moves towards the closest parcel until all parcels are collected.
+  It then visits the closest address of each undelivered parcel until all are
+  delivered.
+*/
+function closestParcelsRobot({place, parcels}, route) {
+  if (route.length === 0) {
```

Comment:
It would make more sense to turn this into a guard clause

Reply:
Turning this into a guard clause would mean duplicating the return statement though. Which is worse than a giant if block right?

Reply:
Nah, the giant if block is WAY worse. It's much more mental overhead because you have to remember what each layer of indentation does. Duplication isn't a big deal, don't let anyone tell you it is.